File: core/src/main/java/com/codeabovelab/tpc/core/doc/DocumentFieldImpl.java
package com.codeabovelab.tpc.core.doc;

import com.codeabovelab.tpc.core.text.TextConsumer;
import com.codeabovelab.tpc.core.text.TextImpl;
import lombok.Data;

/**
 */
@Data
public final class DocumentFieldImpl implements DocumentField {

    @Data
    public static final class Builder implements DocumentField.Builder {
        private String name;
        private String data;

        public Builder name(String name) {
            setName(name);
            return this;
        }

        public Builder data(String data) {
            setData(data);
            return this;
        }

        @Override
        public DocumentFieldImpl build(Document.Builder document) {
            return new DocumentFieldImpl(document.getId(), this);
        }
    }

    private final String id;
    private final String name;
    private final TextImpl data;

    public DocumentFieldImpl(String parentId, Builder b) {
        this.name = b.name;
        this.id = parentId + this.name;
        this.data = new TextImpl(this.id, b.data);
    }

    @Override
    public void read(TextConsumer consumer) {
        consumer.consume(data);
    }
}

File: common/src/main/java/com.codeabovelab.tpc/doc/DocumentFieldImpl.kt
package com.codeabovelab.tpc.doc

import com.codeabovelab.tpc.text.TextConsumer
import com.codeabovelab.tpc.text.TextImpl
import com.codeabovelab.tpc.text.Textual
import com.codeabovelab.tpc.text.TextualUtil

/**
 */
class DocumentFieldImpl private constructor(b: DocumentFieldImpl.Builder) : DocumentField {

    class Builder : DocumentField.Builder<Builder> {
        override var id: String? = null
        override var parent: Textual? = null
        override var childs = ArrayList<Textual.Builder<*>>()
        var data: String? = null

        fun data(data: String?): Builder {
            this.data = data
            return this
        }

        override fun build(): DocumentFieldImpl {
            return DocumentFieldImpl(this)
        }
    }

    override val id: String = b.id!!
    override val parent: Textual? = b.parent!!
    val data: TextImpl
    override val childs: List<Textual> = TextualUtil.buildChilds(this, b)

    init {
        this.data = TextImpl(b.data.orEmpty())
    }

    override fun read(consumer: TextConsumer) {
        consumer(this, data)
    }
}
